### PR TITLE
Add Microsoft static code analyzers

### LINKED ISSUE #2760: [Improvement] Add static code analyzers
I would like to suggest to install [Microsoft.CodeAnalysis.NetAnalyzers](https://github.com/dotnet/roslyn-analyzers) to enhance code quality and also [Microsoft.CodeAnalysis.CSharp.CodeStyle](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/) for styling/formatting rules.

This would help this project to increase the overall quality but also help to enforce certain code style you would like to see.

## Context

The first package would help to prevent certain code smells and future bugs to sneak into the codebase. The second helps to have a defined code style and could help contributors to "code" the way you prefer.

## Possible Implementation

There are a lot of alternative code analyzers but these two are Microsoft actively contributed and are becoming a defacto standard in .NET5+ (can be opt-in/opt-out).

### PR DESCRIPTION
**Pull request by Evangelink**
Fixes #2760

I have used the "default" sets of rules tweaked a bit based on some of the inputs from your resharper PR. This PR may already be a little too big so feel free to let me know if you prefer to split it.

Some split idea:
1. Per analyzer
2. Infra then 1 PR per "rule"

EDIT:
There are a couple of diagnostics being raised as info as they are awaiting for a decision:
- CA1816: Dispose. In a couple of places in the code the `Dispose` pattern is not implemented correctly. I could implement it but there are some open questions: can some classes be marked as `sealed` (in this case there is no need for the pattern). There is one case where `Dispose` is `virtual`, I don't know if I can make the change (causing a potential API break, if that's part of what you consider to be the API of your library).

- CA2219: Do not raise from finally. Is it fine for you to make a change (flags) and throw outside?

- CA1016: Mark assemblies with assembly version.

- CA2211: Non constant fields should not be visible

- CA1018: Specify attribute usage

- CA1806: unused HResult. There is a call to `SetError` in a `Dispose` and the result is ignored. Ignoring HResult can lead to weird behaviors but in the case of a `Dispose` we may consider that to be ok. I don't know enough the code to be able to make such call.

### CODE REVIEW

File: src/GitTools.Testing/Fixtures/RepositoryFixtureBase.cs [hunk 1]
```diff
@@ -18,7 +18,7 @@ protected RepositoryFixtureBase(Func<string, IRepository> repoBuilder)
 
         protected RepositoryFixtureBase(IRepository repository)
         {
-            _sequenceDiagram = new SequenceDiagram();
+            this._sequenceDiagram = new SequenceDiagram();
```
Comment (asbjornu) [inline src/GitTools.Testing/Fixtures/RepositoryFixtureBase.cs:21]: +1 to the use of `this.`, but can you please have the underscore removed as well?

```suggestion
            this.sequenceDiagram = new SequenceDiagram();
```

**Review by asbjornu — CHANGES_REQUESTED** [review]
Besides the failing `dotnet format` test and the remaining `_` in the codebase, this is quite wonderful. Thank you!

The only thing I can think of to add, perhaps in a followup pull request, is `null` guards in all constructors.

File: src/GitVersion.Core.Tests/VersionCalculation/BaseVersionCalculatorTests.cs [hunk 2]
```diff
@@ -18,15 +18,12 @@ public class BaseVersionCalculatorTests : TestBase
...
-            var versionCalculator = GetBaseVersionCalculator(contextBuilder =>
-            {
-                contextBuilder.OverrideServices(services =>
-                {
-                    services.RemoveAll<IVersionStrategy>();
-                    services.AddSingleton<IVersionStrategy>(new V1Strategy(DateTimeOffset.Now));
-                    services.AddSingleton<IVersionStrategy>(new V2Strategy(dateTimeOffset));
-                });
-            });
+            var versionCalculator = GetBaseVersionCalculator(contextBuilder => contextBuilder.OverrideServices(services =>
```
Comment (arturcic) [inline src/GitVersion.Core.Tests/VersionCalculation/BaseVersionCalculatorTests.cs:21]: I'd prefer to keep the curly braces here as it's more readable and aligned

File: src/GitVersion.Core.Tests/VersionCalculation/BaseVersionCalculatorTests.cs [hunk 3]
```diff
@@ -40,15 +37,12 @@ public void UsesWhenFromNextBestMatchIfHighestDoesntHaveWhen()
...
-            var versionCalculator = GetBaseVersionCalculator(contextBuilder =>
-            {
-                contextBuilder.OverrideServices(services =>
-                {
-                    services.RemoveAll<IVersionStrategy>();
-                    services.AddSingleton<IVersionStrategy>(new V1Strategy(when));
-                    services.AddSingleton<IVersionStrategy>(new V2Strategy(null));
-                });
-            });
+            var versionCalculator = GetBaseVersionCalculator(contextBuilder => contextBuilder.OverrideServices(services =>
```
Comment (arturcic) [inline src/GitVersion.Core.Tests/VersionCalculation/BaseVersionCalculatorTests.cs:40]: I'd prefer to keep the curly braces here as it's more readable and aligned

File: src/GitVersion.Core.Tests/VersionCalculation/BaseVersionCalculatorTests.cs [hunk 4]
```diff
@@ -62,15 +56,12 @@ public void UsesWhenFromNextBestMatchIfHighestDoesntHaveWhenReversedOrder()
...
-            var versionCalculator = GetBaseVersionCalculator(contextBuilder =>
-            {
-                contextBuilder.OverrideServices(services =>
-                {
-                    services.RemoveAll<IVersionStrategy>();
-                    services.AddSingleton<IVersionStrategy>(new V1Strategy(null));
-                    services.AddSingleton<IVersionStrategy>(new V2Strategy(when));
-                });
-            });
+            var versionCalculator = GetBaseVersionCalculator(contextBuilder => contextBuilder.OverrideServices(services =>
```
Comment (arturcic) [inline src/GitVersion.Core.Tests/VersionCalculation/BaseVersionCalculatorTests.cs:59]: I'd prefer to keep the curly braces here as it's more readable and aligned

File: src/GitVersion.Core/BuildAgents/GitHubActions.cs [hunk 5]
```diff
@@ -17,19 +17,15 @@ public GitHubActions(IEnvironment environment, ILog log) : base(environment, log
...
+            string.Empty;
 
-        public override string[] GenerateSetParameterMessage(string name, string value)
-        {
+        public override string[] GenerateSetParameterMessage(string name, string value) =>
             // There is no equivalent function in GitHub Actions.
 
-            return new string[0];
-        }
+            System.Array.Empty<string>();
```
Comment (arturcic) [inline src/GitVersion.Core/BuildAgents/GitHubActions.cs:28]: maybe move it on one line?

File: src/GitVersion.LibGit2Sharp/Git/GitRepository.cs [hunk 6]
```diff
@@ -95,72 +95,69 @@ private int GetNumberOfUncommittedChangesInternal()
...
-                }
-                else
-                {
-                    var message = $"Remote tip '{canonicalName}' from remote '{remote.Url}' doesn't look like a valid pull request.";
-                    throw new WarningException(message);
-                }
-            });
-        }
+        public void CreateBranchForPullRequestBranch(AuthenticationInfo auth) => RepositoryExtensions.RunSafe(() =>
+                                                                               {
```
Comment (arturcic) [inline src/GitVersion.LibGit2Sharp/Git/GitRepository.cs:99]: I'd prefer to keep the curly braces here as it's more readable and aligned

File: src/GitVersion.MsBuild.Tests/Helpers/MsBuildTaskFixture.cs [hunk 7]
```diff
@@ -14,47 +14,38 @@ public class MsBuildTaskFixture
...
-        }
+        public void WithEnv(params KeyValuePair<string, string>[] envs) => this.environmentVariables = envs;
 
-        public MsBuildTaskFixtureResult<T> Execute<T>(T task) where T : GitVersionTaskBase
-        {
-            return UsingEnv(() =>
-            {
-                var buildEngine = new MockEngine();
+        public MsBuildTaskFixtureResult<T> Execute<T>(T task) where T : GitVersionTaskBase => UsingEnv(() =>
+                                                                                                        {
```
Comment (arturcic) [inline src/GitVersion.MsBuild.Tests/Helpers/MsBuildTaskFixture.cs:22]: I'd prefer to keep the curly braces here as it's more readable and aligned

File: src/GitVersion.sln [hunk 8]
```diff
@@ -30,9 +31,9 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "GitVersion.MsBuild", "GitVe
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "GitVersion.MsBuild.Tests", "GitVersion.MsBuild.Tests\GitVersion.MsBuild.Tests.csproj", "{B35EF742-95D4-4E6A-B782-BE1C3E5CF9AE}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "GitTools.Testing", "GitTools.Testing\GitTools.Testing.csproj", "{6138640B-5C05-4070-BE32-6B17BEFDF63C}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "GitTools.Testing", "GitTools.Testing\GitTools.Testing.csproj", "{6138640B-5C05-4070-BE32-6B17BEFDF63C}"
```
Comment (arturcic) [inline src/GitVersion.sln:34]: why did this GUID changed?

**Review by arturcic — CHANGES_REQUESTED** [review]

**Comment by arturcic** [comment]
> Besides the failing `dotnet format` test and the remaining `_` in the codebase, this is quite wonderful. Thank you!
> 
> The only thing I can think of to add, perhaps in a followup pull request, is `null` guards in all constructors.

I do agree with removing _.

There are 2 more things to add.
- I'm not a big fan of too long lines so in a follow up PR I think we need to restrict the length of lines to make it more readable
- I noticed in couple of places there are `var => var2 => statements`. I would prefer to have the second pair with curly braces as it looks much more readable.

And as I suggested, please add the nuget packages to Directory.build.props instead of the csprojs

File: src/GitTools.Testing/Fixtures/RepositoryFixtureBase.cs (hunk 1 above)
Comment (Evangelink) [reply src/GitTools.Testing/Fixtures/RepositoryFixtureBase.cs:21]: I will remove them manually but afaik the analyzers don't have a rule to catch this case (nor a code-fix to edit it).

File: src/GitVersion.Core.Tests/Configuration/DefaultConfigFileLocatorTests.cs [hunk 9]
```diff
@@ -191,50 +182,47 @@ public void NoWarnOnCustomYmlFileOutsideRepoPath()
...
-                {
-                    if (log != null) services.AddSingleton(log);
-                    services.AddSingleton(Options.Create(gitVersionOptions));
-                });
-            }
+            private static IServiceProvider GetServiceProvider(GitVersionOptions gitVersionOptions, ILog log = null) => ConfigureServices(services =>
+                                                                                                                                      {
+                                                                                                                                          if (log != null) services.AddSingleton(log);
+                                                                                                                                          services.AddSingleton(Options.Create(gitVersionOptions));
+                                                                                                                                      });
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/GitVersion.Core.Tests/Configuration/DefaultConfigFileLocatorTests.cs:225]: Interesting, seems like the issue comes from the code-fix. I will see if I can create a repro and then create an issue on the analyzer.

File: src/GitVersion.sln (hunk 8 above)
Comment (Evangelink) [reply src/GitVersion.sln:34]: No idea... I have tried to revert but if I reopen in VS it gets updated again. I am wondering if that's linked to the version of VS.

File: src/GitVersion.Core.Tests/VersionCalculation/BaseVersionCalculatorTests.cs (hunk 2 above)
Comment (Evangelink) [reply src/GitVersion.Core.Tests/VersionCalculation/BaseVersionCalculatorTests.cs:21]: Sorry I missed this formatting issue. Just to be sure what's your preference between

1.
```csharp
var versionCalculator = GetBaseVersionCalculator(contextBuilder =>
    contextBuilder.OverrideServices(services =>
    {
        services.RemoveAll<IVersionStrategy>();
        services.AddSingleton<IVersionStrategy>(new V1Strategy(DateTimeOffset.Now));
        services.AddSingleton<IVersionStrategy>(new V2Strategy(dateTimeOffset));
    }));
```

2.
```csharp
var versionCalculator = GetBaseVersionCalculator(contextBuilder =>
{
    contextBuilder.OverrideServices(services =>
    {
        services.RemoveAll<IVersionStrategy>();
        services.AddSingleton<IVersionStrategy>(new V1Strategy(DateTimeOffset.Now));
        services.AddSingleton<IVersionStrategy>(new V2Strategy(dateTimeOffset));
    }
}));
```

In case you go with choice 2, we will probably have to either reduce the severity, prevent preference of expression body for lambdas or use pragam disable for such places.

Comment (arturcic) [reply src/GitVersion.Core.Tests/VersionCalculation/BaseVersionCalculatorTests.cs]: option 1 seems fine if the second expression is on new line and aligned as in the sample

File: src/GitVersion.sln (hunk 8 above)
Comment (arturcic) [reply src/GitVersion.sln:34]: ok I guess not a big issue. we can keep it

File: src/GitTools.Testing/Fixtures/RemoteRepositoryFixture.cs [hunk 10]
```diff
@@ -11,10 +11,7 @@ namespace GitTools.Testing
     public class RemoteRepositoryFixture : RepositoryFixtureBase
     {
         public RemoteRepositoryFixture(Func<string, IRepository> builder)
-            : base(builder)
-        {
-            CreateLocalRepository();
-        }
+            : base(builder) => CreateLocalRepository();
```
Comment (Evangelink) [inline src/GitTools.Testing/Fixtures/RemoteRepositoryFixture.cs:14]: There is no analyzer to help with this kind of formatting. I will make the changes manually, just let me know your preference between:

1. Always break after `=>`
2. Always break before `=>` (that's the one I usually go for but that's purely personal)
3. Break only on long line.

File: src/GitTools.Testing/Fixtures/RepositoryFixtureBase.cs [hunk 11]
```diff
@@ -33,7 +33,7 @@ public string RepositoryPath
 
         public SequenceDiagram SequenceDiagram
         {
-            get { return _sequenceDiagram; }
+            get { return this.sequenceDiagram; }
```
Comment (Evangelink) [inline src/GitTools.Testing/Fixtures/RepositoryFixtureBase.cs:36]: I forgot to handle properties and indexers. Shall I push new changes or do a follow-up PR?

File: src/GitTools.Testing/Fixtures/SequenceDiagram.cs [hunk 12]
```diff
@@ -9,89 +9,74 @@ namespace GitTools.Testing
     /// </summary>
     public class SequenceDiagram
     {
-        readonly Dictionary<string, string> _participants = new Dictionary<string, string>();
-        readonly StringBuilder _diagramBuilder;
+        private readonly Dictionary<string, string> participants = new();
```
Comment (Evangelink) [inline src/GitTools.Testing/Fixtures/SequenceDiagram.cs:12]: Forgot to mention I went ahead to use the new syntax and to favor it but I can change that.

File: src/GitTools.Testing/GitTools.Testing.csproj [hunk 13]
```diff
@@ -4,7 +4,7 @@
         <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
     </PropertyGroup>
     <ItemGroup>
-        <PackageReference Include="LibGit2Sharp" Version="$(PackageVersion_LibGit2Sharp)"/>
-        <PackageReference Include="JetBrains.Annotations" Version="$(PackageVersion_JetBrainsAnnotations)" PrivateAssets="All"/>
+        <PackageReference Include="LibGit2Sharp" Version="$(PackageVersion_LibGit2Sharp)" />
```
Comment (Evangelink) [inline src/GitTools.Testing/GitTools.Testing.csproj:7]: Let me know if you want me to revert these changes (applies for all csproj).

File: src/GitTools.Testing/Fixtures/RepositoryFixtureBase.cs (hunk 11 above)
Comment (arturcic) [reply src/GitTools.Testing/Fixtures/RepositoryFixtureBase.cs:36]: add to this pr

File: src/GitTools.Testing/Fixtures/RemoteRepositoryFixture.cs (hunk 10 above)
Comment (arturcic) [reply src/GitTools.Testing/Fixtures/RemoteRepositoryFixture.cs:14]: 3

File: src/GitTools.Testing/GitTools.Testing.csproj (hunk 13 above)
Comment (arturcic) [reply src/GitTools.Testing/GitTools.Testing.csproj:7]: fine by me

**Comment by Evangelink** [comment]
The test seems to be working fine on my machine (VS + command line). Could you make a test on yours and let me know if that works or fails?

**Review by asbjornu — APPROVED** [review]
All tests green. LGTM!

**Comment by Evangelink** [comment]
@asbjornu Merge done.

**Comment by Evangelink** [comment]
Thank you for the multiple build relaunch @asbjornu 

@arturcic Let us know if you feel we should make more changes here. I am more than happy to provide a follow-up PR (or commit) with any extra changes.

**Comment by arturcic** [comment]
@Evangelink If you don't mind I'll check it tomorrow, having day off today

**Comment by Evangelink** [comment]
Hey sure! Enjoy your day off and speak to you soon.

**Comment by arturcic** [comment]
@Evangelink can you rebase on top of main and solve all conflicts, then we can merge it

**Comment by Evangelink** [comment]
@arturcic Sure! Will work on it now. Any preference between rebase + force push and merge main?

**Comment by arturcic** [comment]
I usually rebase as it's cleaner in terms of history. As long as the branch is not merged into the `main` I always sync the PR branch with main by rebase-ing

**Comment by arturcic** [comment]
@Evangelink can you fix formatting? https://github.com/GitTools/GitVersion/runs/3135376706?check_suite_focus=true

**Comment by Evangelink** [comment]
@arturcic My bad! I will fix that today.

**Comment by asbjornu** [comment]
@arturcic, all green! Do you approve? :)

**Review by arturcic — APPROVED** [review]
